priv/repo/migrations/20180426164341_create_hourly_downloads.exs

```diff
@@ -0,0 +1,56 @@
+defmodule Castle.Repo.Migrations.CreateHourlyDownloads do
+  use Ecto.Migration
+
+  def change do
+    create table(:hourly_downloads, primary_key: false) do
+      add :podcast_id, :integer, null: false
+      add :episode_id, :uuid, null: false
+      add :dtim, :utc_datetime, null: false
+      add :count, :integer, null: false
+    end
+
+    execute """
+    CREATE OR REPLACE FUNCTION create_hourly_downloads_partition() RETURNS trigger AS
+    $$
+      DECLARE
+        partition_start DATE;
+        partition_end DATE;
+        partition TEXT;
+      BEGIN
+        partition_start := DATE_TRUNC('MONTH', NEW.dtim);
+        partition_end := partition_start + INTERVAL '1 MONTH';
+        partition := 'hourly_downloads_' || to_char(NEW.dtim,'YYYYMM');
+        IF NOT EXISTS(SELECT relname FROM pg_class WHERE relname=partition) THEN
+          RAISE NOTICE 'A partition has been created %',partition;
+          EXECUTE 'CREATE TABLE ' || partition || ' (CHECK (dtim >= DATE ''' || partition_start || ''' AND dtim < DATE ''' || partition_end || '''), CONSTRAINT ' || partition || '_uniq UNIQUE (episode_id, dtim)) INHERITS (hourly_downloads);';
+          EXECUTE 'CREATE INDEX ' || partition || '_podcast_id_index ON ' || partition || ' (podcast_id);';
+          EXECUTE 'CREATE INDEX ' || partition || '_episode_id_index ON ' || partition || ' (episode_id);';
+          EXECUTE 'CREATE INDEX ' || partition || '_dtim_index ON ' || partition || ' (dtim);';
+        END IF;
+        EXECUTE 'INSERT INTO ' || partition || ' SELECT(hourly_downloads ' || quote_literal(NEW) || ').* ON CONFLICT ON CONSTRAINT ' || partition || '_uniq DO UPDATE SET (podcast_id, count) = (' || NEW.podcast_id || ', ' || NEW.count || ');';
```

Comment:
This makes inserts into the parent table act as upserts into the partition.

Reply:
I'm noting that it ends up being quite a bit slower, inserting into `hourly_downloads` vs `hourly_downloads_201804`.  ~1s vs ~10s for 60K rows.  I'm guessing that's just the cost of running this function on every row?